Build alarms with explicit name instead of reference and other small CFN fixes

austinbyers opened the pull request:
## Background

Every resource dependency (`!Ref`, `!Sub`, or `!GetAtt`) slows down both Panther deployment and teardown, because CloudFormation has to wait for the first resource to completely finish before moving on to the next.

In our 3 main stacks (core, cloud-security, log-analysis), I noticed we have many references that are really just sharing names:

- Alarm groups (custom resources) are passed the function/queue/table name
- IAM policies for Lambda functions
- Environment variables for Lambda functions

In all of these cases, the reference can be replaced with the explicit resource name; it doesn't matter what order they are created in. You can create alarms, policies, and environment variables even if the associated resource doesn't exist yet.

----

Here's an example of flattening the dependency tree by removing unnecessary dependencies. Today, we have:

- LogProcessorDLQ
   - LogProcessorDLQAlarms
   - LogProcessorQueue
      - LogProcessorQueueAlarms
      - LogProcessorFunction
        - LogProcessorAlarms

But all of the alarms (and the function's IAM role) can be created immediately because we know what the queue and function names will be:

- LogProcessorDLQ
  - LogProcessorQueue
    - LogProcessorFunction
- LogProcessorDLQAlarms
- LogProcessorQueueAlarms
- LogProcessorAlarms

(If the queue weren't an event source for the function, the function itself could also be created at the beginning.) Besides removing one level of the dependency tree, it pushes most resources to the beginning, which allows CFN to parallelize everything more efficiently.

---

What will bother developers is the duplication of the resource names - the same Dynamo table name, for example, now has to be copied to everywhere else in the template that references that table. Personally, I don't think that's a big deal:

- We already have to list the resource name explicitly for "cross-stack" references
- Typos are caught very easily during deployment (functions will fail fast if their permissions/environment variables are not correct)
- Typing `!Ref MyQueue` vs `panther-my-queue` is about the same amount of work and both options are susceptible to copy-pasta bugs
- If the duplication really bugs everyone, we can use Mappings to copy the same value everywhere, like we already do with Timeout and Memory.

## Changes

- Remove all resource references that aren't necessary for dependency ordering
- Simplified some IAM policies, combining related policy statements and removing redundant SQS permissions when a function is invoked from a queue (the Serverless transform will add those permissions automatically)

## Testing

- Deployed to `us-east-1`, app works fine, no CW alarms
- Deploy time is slightly improved for all 3 stacks:

![Deploy Time](https://user-images.githubusercontent.com/3608925/97527627-2446e180-1982-11eb-889b-2548fc5a6422.png)

- Teardown time is about the same:
![Teardown Time](https://user-images.githubusercontent.com/3608925/97527628-2446e180-1982-11eb-8b67-56245cdc7198.png)

These were tested in us-east-1 as fresh deploy/teardowns. Certainly not scientifically rigorous and subject to AWS "weather," but strongly indicative of an overall improvement in deploy time

kostaspap commented on deployments/cloud_security.yml:
```diff
@@ -148,7 +148,7 @@ Resources:
       AlarmTopicArn: !Ref AlarmTopicArn
       CustomResourceVersion: !Ref CustomResourceVersion
-      QueueName: !GetAtt AlertProcessorQueue.QueueName
+      QueueName: panther-alert-processor-queue
```
Interesting. So CW will not fail if you try to setup alarms for a queue that doesn't exist? I assume you have check the custom-resources Lambda for errors during deployment?

austinbyers replied:
> So CW will not fail if you try to setup alarms for a queue that doesn't exist?

Correct - CloudWatch alarms are just thresholds on _metrics_, they don't care if the associated infrastructure exists yet.

 > I assume you have checked the custom-resources Lambda for errors during deployment?

I checked for CW alarms, and there were none, but I should also search the logs as well, good idea. I'll deploy again and check logs to make sure there weren't any warnings or errors I might have missed

kostaspap commented on deployments/cloud_security.yml:
```diff
@@ -1367,7 +1331,7 @@ Resources:
           RESOURCES_API_FQDN: !Sub '${ResourcesApiId}.execute-api.${AWS::Region}.${AWS::URLSuffix}'
           RESOURCES_API_PATH: v1
-          SNAPSHOT_QUEUE_URL: !Ref SnapshotQueue
+          SNAPSHOT_QUEUE_URL: !Ref SnapshotQueue # queue is an event source; needs to exist first
```
I see that `PollerFunction` has a dependency on `SnapshotQueue` declared in line 1339. Doesn't this mean you could also construct the ARN here as you do in other cases?

austinbyers replied:
Yeah, here it doesn't matter whether we do the `!Ref` or the explicit construction because there is already a dependency in the events. I'll go ahead and switch it to be explicit just to be consistent - that way we only use `!Ref` when we really need it

austinbyers replied:
I updated all the environment variables to construct the resource explicitly, even if the same resource has to be referenced as an event source.

Besides being consistent, the event source may actually become a separate resource, in which case this will help the Lambda functions deploy faster

kostaspap reviewed (approved):
LGTM

nhakmiller commented on deployments/cloud_security.yml:
```diff
@@ -1249,7 +1213,7 @@ Resources:
               Action: dynamodb:GetItem
-              Resource: !GetAtt ResourcesTable.Arn
+              Resource: !Sub arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/panther-resources
         - Id: KeyValueStore
```
It's not that big of a deal, but for the rules engine you condensed these four separate inlined policies (resource lookup, KeyValueStore, GetPantherPolicySecrets, GetPantherAnalysisSecrets) into a single inline policy with three statements. For consistency's sake you could do that here as well.

austinbyers replied:
Sure! Good idea. I might do another pass of all the IAM policies in general, we have a bad habit of making a new inline policy for every individual action

austinbyers replied:
Done

austinbyers commented:
I'm running out of time today; since this could introduce subtle bugs and I'll be out next week, I'm going to wait to finish this until I get back

austinbyers commented:
Holding off on merging until I finish removing gateways

austinbyers commented:
This is once again ready for review. I've updated the branch (wow, it's almost been a month already), addressed the feedback, and am doing a fresh deploy to make sure there are no problems.

I'll make sure each page of the web app still works, integration tests pass, and no CloudWatch alarms were triggered during deployment

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -191,7 +191,7 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          TABLE_NAME: !Ref AlertForwarderTable
+          TABLE_NAME: panther-alert-forwarder
```
I assume this breaks the dependency on the ddb table. If the ddb table schema changes and we do an update to a running system, can't this break things?

nhakmiller replied:
Dynamo tables are schema-less, if a lambda refers to new/old columns dynamo will just return nothing if it has nothing there is my understanding. When we start using new columns, we generally try to make it backwards compatible by treating nothing being there as a default situation, but not always possible.

austinbyers replied:
Right, changes to the *schema* would actually require destroying the table, and so we never do schema changes. Adding new columns is managed only in the Lambda functions, not in the Dynamo resource (that's what it means to be noSQL)

However, after chatting offline with @rleighton, he clarified: if we add a new Dynamo index, and update the Lambda functions to use that index, the Lambda update could happen first, and the function would start failing trying to read an index which doesn't exist.

I hadn't thought of that race condition, and there could be others like it. I'm going to go ahead and revert the references in the Lambda functions, but I'll keep the explicit naming in the alarm custom resources. Hopefully, that gives a bit of the extra speed without introducing any new race conditions (the alarm ordering really truly does not matter)

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -288,7 +282,7 @@ Resources:
       CustomResourceVersion: !Ref CustomResourceVersion
       ServiceToken: !Sub arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:panther-cfn-custom-resources
-      TableName: !Ref AlertForwarderTable
+      TableName: panther-alert-forwarder
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
This one is just used for the Dynamo alarm group, no race condition

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -561,7 +539,7 @@ Resources:
       Environment:
         Variables:
-          COMPLIANCE_TABLE: !Ref ComplianceTable
+          COMPLIANCE_TABLE: panther-compliance
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Will revert

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -671,8 +647,8 @@ Resources:
           DEBUG: !Ref Debug
-          SQS_QUEUE_URL: !Ref RemediationQueue
-          REMEDIATION_LAMBDA_ARN: !GetAtt RemediationFunction.Arn
+          SQS_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-remediation-queue
```
Could this introduce a race during deployment?

austinbyers replied:
Not likely, but reverting change

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -699,14 +675,14 @@ Resources:
                 - sqs:SendMessage
                 - sqs:SendMessageBatch
-              Resource: !GetAtt RemediationQueue.Arn
+              Resource: !Sub arn:${AWS::Partition}:sqs:${AWS::Region}:${AWS::AccountId}:panther-remediation-queue
```
Another race possible? Lambda executes and no q?

austinbyers replied:
Reverted

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -927,8 +897,8 @@ Resources:
-          RESOURCES_TABLE: !Ref ResourcesTable
+          RESOURCES_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-resources-queue
+          RESOURCES_TABLE: panther-resources
       FunctionName: panther-resources-api
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -927,8 +897,8 @@ Resources:
           DEBUG: !Ref Debug
-          RESOURCES_QUEUE_URL: !Ref ResourcesQueue
-          RESOURCES_TABLE: !Ref ResourcesTable
+          RESOURCES_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-resources-queue
```
Possible execution race?

austinbyers replied:
Reverted

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -1121,21 +1091,15 @@ Resources:
           Version: 2012-10-17
           Statement:
-            - Effect: Allow
-              Action:
```
No longer used?

austinbyers replied:
Ah, no - when you have an event source, the Serverless transform automatically adds the necessary permissions to use it.

Since SQS is feeding this function, we don't need a separate policy to allow reading SQS messages - one will be added automatically

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -1308,7 +1262,7 @@ Resources:
           AUDIT_ROLE_NAME: !Sub PantherAuditRole-${AWS::Region}
           DEBUG: !Ref Debug
-          SNAPSHOT_QUEUE_URL: !Ref SnapshotQueue
+          SNAPSHOT_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-snapshot-queue
```
Race on initial deploy possible?

austinbyers replied:
No, because the queue is still referenced in the event-source mapping. Reverting nontheless

rleighton commented on deployments/cloud_security.yml:
```diff
@@ -1398,7 +1352,7 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          SNAPSHOT_POLLERS_QUEUE_URL: !Ref SnapshotQueue
+          SNAPSHOT_POLLERS_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-snapshot-queue
```
Race on deploy possible?

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -264,17 +264,17 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          ORG_TABLE_NAME: !Ref OrganizationTable
+          ORG_TABLE_NAME: panther-organization
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -319,7 +319,7 @@ Resources:
           RULES_ENGINE: panther-rules-engine
           RESOURCE_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-resources-queue
-          TABLE: !Ref AnalysisTable
+          TABLE: panther-analysis
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -489,7 +485,7 @@ Resources:
           DEBUG: !Ref Debug
           KEY_ID: !Ref OutputsKeyId
-          OUTPUTS_TABLE_NAME: !Ref OutputsTable
+          OUTPUTS_TABLE_NAME: panther-outputs
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -609,7 +605,7 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          ALERT_QUEUE_URL: !Ref AlertQueue
+          ALERT_QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-alerts-queue
```
Race on deploy?

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -759,7 +746,7 @@ Resources:
           LOG_PROCESSOR_QUEUE_ARN: !Sub arn:${AWS::Partition}:sqs:${AWS::Region}:${AWS::AccountId}:panther-input-data-notifications-queue
           SNAPSHOT_POLLERS_QUEUE_URL: !Sub https://sqs.${AWS::Region}.amazonaws.com/${AWS::AccountId}/panther-snapshot-queue
-          TABLE_NAME: !Ref IntegrationsTable
+          TABLE_NAME: panther-source-integrations
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/core.yml:
```diff
@@ -1233,17 +1202,17 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          LOG_TYPES_TABLE_NAME: !Ref LogTypesTable
+          LOG_TYPES_TABLE_NAME: panther-logtypes
```
Not sure about losing dependencies on ddb tables ... see previous comment.

austinbyers replied:
Reverting

rleighton commented on deployments/log_analysis.yml:
```diff
@@ -184,7 +183,7 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          ALERTS_TABLE_NAME: !Ref LogAlertsTable
+          ALERTS_TABLE_NAME: panther-log-alert-info
```
Not sure about losing dependencies on ddb tables ... see previous comment.

rleighton commented on deployments/log_analysis.yml:
```diff
@@ -312,7 +310,7 @@ Resources:
         Variables:
           DEBUG: !Ref Debug
-          ALERTS_TABLE: !Ref LogAlertsTable
+          ALERTS_TABLE: panther-log-alert-info
```
Not sure about losing dependencies on ddb tables ... see previous comment.

rleighton commented on deployments/log_analysis.yml:
```diff
@@ -667,23 +658,23 @@ Resources:
           ATHENA_WORKGROUP: !Ref AthenaWorkGroup
           DEBUG: !Ref Debug
-          QUEUE_URL: !Ref UpdaterQueue
+          QUEUE_URL: !Sub https://sqs.${AWS::Region}.${AWS::URLSuffix}/${AWS::AccountId}/panther-datacatalog-updater-queue
```
Race on initial deploy?

rleighton commented on deployments/log_analysis.yml:
```diff
@@ -908,7 +892,7 @@ Resources:
           S3_BUCKET: !Ref ProcessedDataBucket
           NOTIFICATIONS_TOPIC: !Ref ProcessedDataTopicArn
-          ALERTS_DEDUP_TABLE: !Ref AlertsDedup
+          ALERTS_DEDUP_TABLE: panther-log-alert-dedup
```
Not sure about losing dependencies on ddb tables ... see previous comment.

rleighton commented on deployments/log_analysis.yml:
```diff
@@ -1106,15 +1068,15 @@ Resources:
           DEBUG: !Ref Debug
-          STREAM_NAME: !Ref MessageForwarderFirehose
-      Tracing: !If [TracingEnabled, !Ref TracingMode, !Ref 'AWS::NoValue']
+          STREAM_NAME: panther-forwarder-firehose
```
Race on initial deploy?

rleighton reviewed (changes requested):
Nice clean up! 

I have concerns about losing dependencies of the lambda functions on tables and queues. Couldn't there be cases when a function runs soon after an initial  deploy but a table/queue does not exist? This depends on each function but we did not need to worry about it when we had true dependencies.

Updates of ddb table schema/indexes are another concern when deploying over a running system. Now a lambda could read a ddb table but perhaps it has not been changed to the new version of the schema with an expected added column or index. If we kept dependencies then it is not a concern.

Thoughts?

austinbyers commented:
This wasn't a high-priority PR, more of an experiment I played with late one night :) Reverting changes to the Lambda functions, no need to introduce new race conditions just for the sake of slightly faster deploys

But we have hundreds of alarms and those are still building with the explicit resource name. That change alone should be enough to shave a few seconds off of each stack
